pmd-java/src/test/java/net/sourceforge/pmd/lang/java/oom/DataStructureTest.java

```diff
@@ -30,13 +46,39 @@
  */
 public class DataStructureTest extends ParserTst {
 
+    MetricKey<ClassMetric> classMetricKey = new MetricKey<ClassMetric>() {
+        @Override
+        public String name() {
+            return null;
+        }
+
+
+        @Override
+        public ClassMetric getCalculator() {
+            return new RandomMetric();
```

Comment:
That's a nice idea! It's much cleaner than the current anonymous class strategy. 

Hmm well bottom line, keys should implement hashcode, which is why it's convenient to use enums. They're not used directly to index memoization maps, they're incorporated into a `ParameterizedMetricKey` (metrickey + metricversion). 

As parameterized metric keys are instance controlled, they can use identity based equality, but in order to find the right instance of parameterized metric key, we use the hashcode of the key (see `ParameterizedMetricKey#code(MetricKey, MetricVersion)`). So I guess keys need not even implement equals, even though it's a good idea